First stab at migrating to CometBFT-rs

adizere opened the pull request:
Closes #189, #194

adizere committed:
First stab at migrating to CometBFT-rs

adizere commented on tools/proto-compiler/src/cmd/compile.rs:
```diff
@@ -147,7 +147,7 @@ impl CompileCmd {
             .out_dir(out_dir)
             .file_descriptor_set_path(out_dir.join("proto_descriptor.bin"))
-            .extern_path(".tendermint", "::tendermint_proto")
+            .extern_path(".tendermint", "::cometbft_proto")
```
Simply doing this change is not enough. We need to select between the different options within `cometbft_proto`.

```
error[E0412]: cannot find type `SignedHeader` in module `cometbft_proto::types`
   --> src/prost/ibc.lightclients.tendermint.v1.rs:128:72
    |
128 |     pub signed_header: ::core::option::Option<::cometbft_proto::types::SignedHeader>,
    |                                                                        ^^^^^^^^^^^^ not found in `cometbft_proto::types`
    |
help: consider importing one of these items
   --> src/lib.rs:373:17
    |
373 +                 use cometbft_proto::types::v1::SignedHeader;
   --> |src/lib.rs:373:17
    |
373 +                 use cometbft_proto::types::v1beta1::SignedHeader;
```

cc @mzabaluev any thoughts?

mzabaluev replied:
It seems ibc-proto is still using the `tendermint.*` packages, not `cometbft.*`. To be able to use the modules from `cometbft_proto`, please switch to importing `cometbft` packages on the protobuf level. There you'll need to specify the versioned paths to proto packages you intend to be using with your revision of IBC.

mzabaluev replied:
Trouble is, the `tendermint` protos are still bundled with the SDK repository checkout and they may be outdated.
So this may be blocked on the SDK updating their proto files to depend on `cometbft`, preferably as a buf module.

mzabaluev replied:
On the other hand, the sole reason for the SDK checkout seems to be to import "tendermint/crypto/proof.proto". So I'll try replacing that checkout completely with cometbft protos, let's see if difficulties arise.

adizere replied:
> I'll try replacing that checkout completely with cometbft protos, let's see if difficulties arise.

let's do that

if that doesn't work, let's open a PR on the SDK

mzabaluev replied:
> the sole reason for the SDK checkout seems to be to import "tendermint/crypto/proof.proto".

That wasn't true, of course: cosmos and ICS packages are compiled as part of this library, and `tendermint` is used all over the place in those protos. So this is blocked on upstream protos converting to `cometbft` first.

mzabaluev committed:
proto-compiler: import CometBFT protos

A work in progress to use the cometbft proto module from BSF
instead of the tendermint.* protos in cosmos-sdk.

mzabaluev committed:
Switch to cometbft.crypto.v1 in stride.interchainquery.v1

mzabaluev committed:
Merge branch 'main' into migrate-cometbft

mzabaluev committed:
Don't use cometbft checkout to build protos

mzabaluev committed:
Remap tendermint.* protos to cometbft_proto modules

mzabaluev committed:
Update cosmos-sdk checkout to 0.50.4

This brings cosmos protos up to date with tendermint 0.38, which is the version included by the proto files from the ibc-go checkout.

mzabaluev committed:
Restore temp directory cleanup in sync-protobuf.sh

mzabaluev committed:
Merge branch 'main' into migrate-cometbft

mzabaluev committed:
proto-compiler: fix fmt error

mzabaluev committed:
Remove generated files for protos not in cosmos-sdk 0.50

mzabaluev committed:
Remap some abci protos to the correct cometbft_proto mod

mzabaluev committed:
proto-compiler: fix a fresh fmt error

adizere commented on tools/proto-compiler/src/cmd/compile.rs:
```diff
@@ -142,19 +129,33 @@ impl CompileCmd {
+                "::cometbft_proto::abci::v1beta1::ValidatorUpdate",
             )
-            .extern_path(".tendermint", "::tendermint_proto")
+            .extern_path(".tendermint.abci", "::cometbft_proto::abci::v1beta3")
```
nice, this seems more straightforward than I expected!
